### ISSUE TITLE
Request xrootd protocol as defualt for accessURL

### ISSUE THREAD

**Issue by mesmith75** [ganga-devs/ganga#751]
The accessURL command needs to be changed to get a URL with the xrootd protocol as default:

```
dirac-dms-lfn-accessURL /lhcb/LHCb/Collision11/BHADRON.MDST/00024034/0002/00024034_00024451_1.bhadron.mdst --SE CNAF-DST --Protocol xroot
```

I will add this in.

**Comment by mesmith75** [comment]
Unfortunately the Dirac API doesn't include options for
```
getAccessURL(lfn, storageElement, printOutput=False)
```
Should we change this to using the dirac-dms-lfn-accessURL command?

**Comment by rob-c** [comment]
My reaction is that it would be more useful to see what this script does behind the scenes and use the functions there in the standard wrapper. Changing to use the script directly is a potentially a relatively big undertaking for potentially little gain imo.
Is the behaviour here the same between DIRAC and LHCbDIRAC? (On the face of it I would imagine that it is)

**Comment by mesmith75** [comment]
@rob-c Yes that would make more sense. The relevant part is in the DataManagerSystem of Dirac:
 DIRAC/DataManagementSystem/Client/DataManager.py
https://github.com/DIRACGrid/DIRAC/blob/366315622670082e3e5492769ad52be6d375f3cd/DataManagementSystem/Client/DataManager.py L1660
The call from the API just leaves the protocol argument out.

**Comment by rob-c** [comment]
Where is the accessURL used atm in the codebase? I suspect that setting the protocol to be `xrootd` is not what is preferred for the GridPP instance so I think we would need to make some configurable or wrapper for LHCbDirac calls which do things slightly differently there.

**Comment by mesmith75** [comment]
At the moment just for DiracFiles, although we want to extend that to a static function and jobs with the Dirac backend.
I see your point about GridPP. I think the only issue is really the CNAF-USER SE which as a default doesn't use a protocol that allows outside access.

**Comment by rob-c** [comment]
@mesmith75 Is there a sensible failure associated with the case of `CNAF-USER`? If not we should probably add some sort of mechanism to allow a user to get a file over `xrootd` from another SE unless the user has specifically requested `CNAF` for better or worse.
This quickly sounds like a non-trivial but useful feature to have automated :+1:

**Comment by mesmith75** [comment]
@rob-c It is not really failure. You get back a valid URL so you don't know you can't use it until you try to open it. I guess it is noticeable in the path you get. These are user jobs so usually they only have 1 replica the location of which the user doesn't think about (unless they start banning SEs).

Maybe it would just be best to add in a few lines to make the necessary changes for files at CNAF and leave the rest as is.

As an enhancement it might be nice if the user can specify the protocol themselves. I don't know what the relative merits are beyond different sites having differing defaults. Is it really an issue to get xrootd paths from sites where that is not the default? It still works fine as far as I can make out.

**Comment by rob-c** [comment]
If we're going down the route of blacklisting `CNAF` can we add a configurable list somewhere which would allow users to add/remove remote SE based upon grid related activity. (We can default this to `['CNAF-USER']` within the LHCb config.
If we're going to add some way for LHCb users to have xrootd by default then I think it also makes sense to add an option to the function to allow then to change this if required although this is likely a secondary feature to implement.

**Comment by mesmith75** [comment]
I would rather not start blacklisting sites if we don't like how they are set up. I'll change the function to request the xrootd by default for all sites and add an option for the user to choose their own if they wish. That should be straightforward.

I kind of like the idea of adding removing SEs based on grid activity but I would have thought that Dirac sets the priorities of SEs itself based on that criteria so that would be rather redundant?

**Comment by rob-c** [comment]
There is no other way to disable remote access at CNAF based upon what you've said, I think maintaining a list of sites which don't allow remote access isn't a bad idea. If a user doesn't request a site and they randomly get given a URL for CNAF from a replica there it's going to cause problems imo.

I'd rather not trust being able to poll DIRAC for the order of preference (there is no exposed way of doing this) if it's much easier to store the knowledge within Ganga that SiteX doesn't work for remote access.
The DIRAC API doesn't have any easy way of telling us this and imo something should be setup to fail when requesting an xrootd URL from CNAF, but I'm sure there is a good reason they don't do this. Or the site should be marked as not available or accessible internally so the replica isn't exposed.

We're doing something different here to the internal DIRAC logic of site matching and setting up a job to run so I don't think we can rely on DIRAC

**Comment by mesmith75** [comment]
Ah no, the CNAF does allow remote access - you can request the xrootd URL and it works fine. It is just the default it returns is not for remote, you just need to specify the protocol.

**Comment by rob-c** [comment]
OK, then I suspect we can get away with just using the protocol correctly but having it a safe default in GangaDirac which is set to `xrootd` in GangaLHCb

**Comment by mesmith75** [comment]
Closed by #754
